# 谜题 37：极端不可思议  

## 背景

 捕获异常类型。

## 代码&陷阱

下面的三个程序每一个都会打印些什么？不要假设它们都可以通过编译。

### 1、第一个程序

它看起来应该是可以编译的：try 子句执行 I/O，并且 catch 子句捕获 IOException 异常。

```java
import java.io.IOException;

public class Demo1 {

    public static void main(String[] args) {
        try {
            System.out.println("Hello world");
        } catch(IOException e) {
            System.out.println("I've never seen println fail!");
        }
    }
}

```

这个程序不能编译，因为 println 方法没有声明会抛出任何被检查异常，而IOException 却正是一个被检查异常。Demo1，展示了被检查异常的一个基本原则。

语言规范中描述道：**如果一个 catch 子句要捕获一个类型为 E 的被检查异常，而其相对应的 try 子句不能抛出 E 的某种子类型的异常，那么这就是一个编译期错误**。 

### 2、第二个程序 

Demo2，看起来应该是不可以编译的，但是它却可以。  

```java

public class Demo2 {
    public static void main(String[] args) {
        try {
            // If you have nothing nice to say, say nothing
        } catch(Exception e) {
            System.out.println("This can't happen");
        }
    }
}
```

它之所以可以编译，是因为它唯一的 catch 子句检查了 Exception，**捕获 Exception 或 Throwble 的 catch 子句是合法的，不管与其相对应的 try 子句的内容为何**。

尽管 Demo2是一个合法的程序，但是 catch 子句的内容永远的不会被执行，这个程序什么都不会打印。  

### 3、第三个程序

Demo3，看起来它也不能编译。  

```java
interface TypeA {
    void f() throws CloneNotSupportedException;
}

interface TypeB {
    void f() throws InterruptedException;
}

interface TypeC extends TypeA, TypeB {
}

public class Demo3 implements TypeC {
    public void f() {
        System.out.println("Hello world");
    }
    
    public static void main(String[] args) {
        TypeC t = new Demo3();
        t.f();
    }
}
```

方法 f 在 TypeA接口中声明要抛出被检查异常 CloneNotSupportedException，并且在 TypeB接口中声明要抛出被检查异常 InterruptedException。

TypeC接口继承了 TypeA和 TypeB，因此，看起来在静态类型为 TypeC的对象上调用方法 f时，有潜在可能会抛出这些异常。

Demo3的 main 方法在静态类型为 TypeC的对象上调用了方法 f，但它对 CloneNotSupportedException 和 InterruptedExceptioin 并没有作这些处理。那么，为什么这个程序可以编译呢？    

**规则：一个方法可以抛出的被检查异常集合，是它所适用的所有类型声明要抛出的被检查异常集合的交集，而不是合集**。因此，静态类
型为 TypeC的对象上的 f 方法根本就不能抛出任何被检查异常。因此，Demo3可以毫无错误地通过编译，并且打印 Hello world。  

## 解决办法或规则

总之，第一个程序说明了一项基本要求，即**对于捕获被检查异常的 catch 子句，只有在相应的 try 子句可以抛出这些异常时才被允许**。

第二个程序说明了这项要求不会应用到的冷僻案例。

第三个程序说明了**多个继承而来的 throws 子句的交集，将减少而不是增加方法允许抛出的异常数量**。  